js/hw_2_js_functions_state/src/game_of_snake/game_of_snake.js

```diff
@@ -1,3 +1,440 @@
+});
+
+engine.addGlobalReducer((state) => {
+  if (state.gameTimer < 1000) {
+    return {
+      ...state,
+      gameLost: true,
+      snakeCoords: [],
+      dishIndexes: [],
+    };
+  }
+
+  return state;
+});
+
+engine.addSignalReducer('startTimer', (state) => {
+  return {
+    ...state,
+    gameTimer: state.gameTimer - 100,
+  };
+});
+
+engine.addSignalReducer('moveRight', (state) => {
+  if (state.direction !== 'moveLeft') {
+    return {
+      ...state,
+      direction: 'moveRight',
+    };
+  }
+
+  return state;
+});
+
+engine.addSignalReducer('moveLeft', (state) => {
+  if (state.direction !== 'moveRight') {
+    return {
+      ...state,
+      direction: 'moveLeft',
+    };
+  }
+
+  return state;
+});
+
+engine.addSignalReducer('moveDown', (state) => {
+  if (state.direction !== 'moveUp') {
+    return {
+      ...state,
+      direction: 'moveDown',
+    };
+  }
+
+  return state;
+});
+
+engine.addSignalReducer('moveUp', (state) => {
+  if (state.direction !== 'moveDown') {
+    return {
+      ...state,
+      direction: 'moveUp',
+    };
+  }
+
+  return state;
+});
+
+engine.addSignalReducer('pauseGame', (state) => {
+  return {
+    ...state,
+    gameOn: !state.gameOn,
+  };
+});
+
+engine.addSignalReducer('moveSnake', (state) => {
+  const prevSnakeHead = { ...state.snakeHead };
+  let x = prevSnakeHead.x;
+  let y = prevSnakeHead.y;
+
+  switch (state.direction) {
+    case 'moveDown':
+      y++;
+      break;
+    case 'moveUp':
+      y--;
+      break;
+    case 'moveRight':
+      x++;
+      break;
+    case 'moveLeft':
+      x--;
+      break;
+
+    default:
+      break;
+  }
+
+  const nextSnakeHead = {
+    x,
+    y,
+  };
+
+  const minBorderCoords = 0;
+  const maxBorderCoords = 26;
+  const isSnakeHitBorders =
+    x === maxBorderCoords ||
+    y === maxBorderCoords ||
+    x === minBorderCoords ||
+    y === minBorderCoords;
+  const isSnakeHitBody = state.snakeCoords.includes(getIndex(x, y));
+
+  if (isSnakeHitBody || isSnakeHitBorders) {
+    return {
+      ...state,
+      gameLost: true,
+      snakeCoords: [],
+      dishIndexes: [],
+    };
+  }
+
+  const catchDishPoint = 1;
+  const snakeLength = state.snakeCoords.length;
+  const snakeHeadCoordinates = [...state.snakeCoords][snakeLength - 1];
+  const prevStateDishIndexes = [...state.dishIndexes];
+  const indexDishToRemove = prevStateDishIndexes.indexOf(snakeHeadCoordinates);
+  const newData = state.dishIndexes.includes(snakeHeadCoordinates)
+    ? [...state.snakeCoords]
+    : [...state.snakeCoords].slice(1, snakeLength);
+
+  if (indexDishToRemove > -1) {
+    prevStateDishIndexes.splice(indexDishToRemove, 1);
+  }
+
+  return {
+    ...state,
+    snakeHead: nextSnakeHead,
+    currentScore: state.dishIndexes.includes(snakeHeadCoordinates)
+      ? state.currentScore + catchDishPoint
+      : state.currentScore,
+    dishIndexes: state.dishIndexes.includes(snakeHeadCoordinates)
+      ? prevStateDishIndexes
```

Comment:
Doesn't sound right: both `state.dishIndexes` and `prevStateDishIndexes` refer to **previous** state.
Naming should be changed.